### ISSUE TITLE
Runnaway `ecm`

### ISSUE THREAD

**Issue by calebTree** [RsaCtfTool/RsaCtfTool#381]
The use of `RsaCtfTool/sage/ecm.sage.py` continues after `RsaCtfTool/RsaCtfTool.py` ends.

```
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
kali       23162  0.8  0.3 105400 53388 pts/0    S+   19:37   0:00 python3 ./RsaCtfTool.py --publickey ../../Desktop/metro.pub --private --attack ecm
kali       23164  0.0  0.0   2728   924 pts/0    S+   19:37   0:00 /bin/sh /home/kali/apps/sage-9.7/src/bin/sage-python /home/kali/apps/sage-9.7/src/bin/sage-run /home/kali/apps/RsaCtfTool/lib/../sage/ecm.sage 267
kali       23271  0.0  0.0   2728   900 pts/0    S+   19:37   0:00 /bin/sh /home/kali/apps/sage-9.7/src/bin/sage-python /home/kali/apps/sage-9.7/src/bin/sage-cleaner
kali       23272  0.0  0.0   2728   900 pts/0    S+   19:37   0:00 /bin/sh /home/kali/apps/sage-9.7/src/bin/sage-python /home/kali/apps/RsaCtfTool/lib/../sage/ecm.sage.py 267041360701900212685306777951059858842633
kali       23273  0.1  0.0  14052 10972 pts/0    S+   19:37   0:00 python3 /home/kali/apps/sage-9.7/src/bin/sage-cleaner
kali       23389  3.1  1.1 1400948 183256 pts/0  S+   19:37   0:00 python3 /home/kali/apps/RsaCtfTool/lib/../sage/ecm.sage.py 267041360701900212685306777951059858842633615366228339347144268884267986659635180346561
kali       23429 99.8  0.0   9264  7428 pts/0    R+   19:37   0:19 ecm -c 1000000000 -I 1 -one 2000
kali       23511  0.0  0.0   9872  3268 pts/1    R+   19:37   0:00 ps aux
```

**Comment by daedalus** [comment]
If you can trace the root cause of this issue it will be much appreciated.

**Comment by MarcoMeinardi** [comment]
I think the problem is that the child of `ecm.py` (`ecm.sage.py`) spawn a new child process (`ecm`), thus, it doesn't get killed. I'm not entirely sure about that, because at least the `ecm.sage.py` process should be killed, but it is not.
Maybe something like [this](https://stackoverflow.com/questions/4789837/how-to-terminate-a-python-subprocess-launched-with-shell-true) will work.

**Comment by calebTree** [comment]
> I think the problem is that the child of `ecm.py` (`ecm.sage.py`) spawn a new child process (`ecm`), thus, it doesn't get killed. I'm not entirely sure about that, because at least the `ecm.sage.py` process should be killed, but it is not. Maybe something like [this](https://stackoverflow.com/questions/4789837/how-to-terminate-a-python-subprocess-launched-with-shell-true) will work.

Yes, that was basically it. I do not have documentation on where and when `sage` creates child processes however it is apparent that is the case. I started with this well written blog post illustrating the issue:
- [Kill a Python subprocess and its children when a timeout is reached](https://alexandra-zaharia.github.io/posts/kill-subprocess-and-its-children-on-timeout-python/)

Then I adapted what my understanding should be a reasonable cross-platform method of terminating a process group from this StackOverflow:
- [subprocess: deleting child processes in Windows](https://stackoverflow.com/a/4229404/12648588)

I decided to work on a PR which implements the more verbose `subprocess.Popen()`, `wait()`, `communicate()` interface. Instead of `subprocess.check_output()`.

**Comment by calebTree** [comment]
> If you can trace the root cause of this issue it will be much appreciated.

@daedalus
How it looks ot me, `sage` does spawn child processes that are not terminated using the `subprocess.check_output()` interface.

**Comment by daedalus** [comment]
fixed in  #384
